scheduler/src/cook/mesos/scheduler.clj

```diff
@@ -279,9 +279,121 @@
   (-> (d/entity db [:instance/task-id task-id])
       :db/id))
 
+(histograms/defhistogram [cook-mesos scheduler progress-aggregator-pending-states])
+(meters/defmeter [cook-mesos scheduler progress-aggregator-message-rate])
+(meters/defmeter [cook-mesos scheduler progress-aggregator-drop-rate])
+(counters/defcounter [cook-mesos scheduler progress-aggregator-drop-count])
+
+(defn progress-update-aggregator
+  "Launches a long running go block that triggers publishing the latest aggregated instance-id->progress-state
+   whenever there is a message on the progress-aggregator-trigger-chan.
+   It drops messages if the progress-aggregator-chan queue is larger than pending-progress-threshold or the aggregated
+   state has more than pending-progress-threshold different entries.
+   It returns a map containing progress-aggregator-chan and query-state-chan.
+   progress-aggregator-chan can be used to send progress-state messages to the aggregator.
+   query-state-chan can be used to query the state of the aggregator."
+  [pending-progress-threshold progress-aggregator-trigger-chan progress-state-chan]
+  (log/info "Starting progress update aggregator")
+  (let [progress-aggregator-chan (async/chan (async/sliding-buffer pending-progress-threshold))
+        query-state-chan (async/chan (async/sliding-buffer 1))]
+    (async/go
+      (loop [instance-id->progress-state {}]
+        (let [[data chan] (async/alts! [progress-aggregator-trigger-chan progress-aggregator-chan query-state-chan]
+                                       :priority true)]
+          (condp = chan
+            progress-aggregator-trigger-chan
```

Comment:
I can see how we would use transducers to update `instance-id->progress-state` as we receive data on `progress-aggregator-chan` by piping the latest value to a sliding buffer chan. However, I'm not sure how I can atomically clear out the portion of the map passed on to `progress-state-chan` read from this new chan as `progress-aggregator-chan` can (1) receive intermediate progress update and (2) it itself has a sliding buffer and can drop messages.